modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modeling
+
+import (
+	"container/list"
+	"fmt"
+	"math"
+)
+
+// LocalSliceAlgorithm 'borrows' and 'rents' endpoints from other zones to make
+// the local EndpointSlice balanced with the incoming traffic (number of nodes
+// distribution)
+type LocalSliceAlgorithm struct{}
+
+type endpointDeviation struct {
+	name      string
+	deviation float64
+}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy
+func (alg LocalSliceAlgorithm) CreateSliceGroups(region regionInfo) (map[string]EndpointSliceGroup, error) {
+	if region.zoneDetails == nil {
+		return nil, fmt.Errorf("can't create EndpointSlices with 0 number of zone")
+	}
+	sliceGroups := map[string]EndpointSliceGroup{}
+	// stack stores 'endpointDeviation' -- the info of how should number of
+	// endpoints in a zone be adjusted to match the ratio balance
+	endpointStack := list.New()
+	for name, zone := range region.zoneDetails {
+		var localGroup EndpointSliceGroup
+		localGroup.Label = name
+		localGroup.Weights = map[string]float64{name: 1.0}
+		localGroup.Composition = map[string]int{}
+
+		deviation := math.Round(zone.nodesRatio*float64(region.totalEndpoints) - float64(zone.Endpoints))
```

Comment:
Also could use a comment here to explain the math here.

Reply:
done